Principal Component Analysis Aggregation

nknize opened the pull request:
This PR adds a new aggregation for computing principal component analysis over an array of numeric fields. The input is an array of field names:
```
"myPCA" : {
    "pca" : {
        "fields" : ["income", "savings", "age", "education"]
    }
}
```
The output is the list of field names (could be in a different order than provided at input) along with the eigen values, and eigen vectors needed to either transform the original data (into a reduced dimensional set), or determine which fields are of significance.

```
"myPCA" : {
    "fields" : ["income", "education", "age", "savings"],
    "pc" : [
        {
            "eigenvalue" : 3.5476,
            "eigenvector" : [0.314, 0.145, -0.676, 0.494]
        },
        {
            "eigenvalue" : 2.1320,
            "eigenvector" : [0.237, 0.444, -0.401, -0.357]
        },
        {
            "eigenvalue" : 1.0447,
            "eigenvector" : [0.484, -0.135, -0.004, -0.487]
        },
        {
            "eigenvalue" : 0.1665,
            "eigenvector" : [0.404, 0.219, 0.366, 0.568]
        }
    ]
}
```

### Implementation Notes:

1. Basic licensed
2. Included in the Analytics x-pack plugin
3. Implemented on top of matrix_stats OSS module; some of Matrix Stats classes are abstracted
4. Adds [ejml](http://ejml.org/) linear algebra library dependency (Apache licensed)
5. Default computation is achieved by running eigen decomposition on the correlation matrix
6. Option `use_covariance` enables user to run eigen decomposition on the covariance matrix
7. Docs are provided; `pca` documentation is under the `Aggregations` section and the x-pack label is applied

nknize committed:
PCA Aggregation

This commit adds a new aggregation for computing principal component analysis over an array of numeric fields. The input is an array of field names:

"myPCA" : {
    "pca" : {
        "fields" : ["income", "savings", "age", "education"]
    }
}

The output is the list of field names (could be in a different order than provided at input) along with the eigen values, and eigen vectors needed to either transform the original data (into a reduced dimensional set), or determine which fields are of significance.

"myPCA" : {
    "fields" : ["income", "education", "age", "savings"],
    "pc" : [
        {
            "eigenvalue" : 3.5476,
            "eigenvector" : [0.314, 0.145, -0.676, 0.494]
        },
        {
            "eigenvalue" : 2.1320,
            "eigenvector" : [0.237, 0.444, -0.401, -0.357]
        },
        {
            "eigenvalue" : 1.0447,
            "eigenvector" : [0.484, -0.135, -0.004, -0.487]
        },
        {
            "eigenvalue" : 0.1665,
            "eigenvector" : [0.404, 0.219, 0.366, 0.568]
        }
    ]
}

nknize committed:
add licenses, fix headers, fix checkstyle, pass precommit

nknize committed:
finish toXContent implementation

nknize committed:
Switch to using the correlation matrix for computing eigen decomposition and add a new 'useCovariance' aggregation option to choose between covariance or correlation.

nknize committed:
add PCA documentation to ML API reference docs

nknize committed:
minor doc cleanup; to relocate in another commit

nknize committed:
fix NamedXContentProvider service resource

nknize committed:
move pca aggregation documentation from ml to Aggregations section

elasticmachine commented:
Pinging @elastic/ml-core

elasticmachine commented:
Pinging @elastic/es-analytics-geo

nknize committed:
fix doc failure

droberts195 commented on docs/reference/aggregations/matrix/pca-aggregation.asciidoc:
```diff
@@ -0,0 +1,158 @@
+[role="xpack"]
+[testenv="platinum"]
```
Should it be `[testenv="basic"]` since the PR description says this feature is basic license?

droberts195 reviewed (commented):
One side-effect of putting this in the `ml` plugin that it's worth being aware of is that if people disable `ml` for any reason they'll lose access to this aggregation.  Some users set `xpack.ml.enabled: false` for various hardware or OS incompatibilities caused by ML's native code components, for example see https://discuss.elastic.co/t/elasticsearch-7-1-1-failed-to-create-native-process-factories-for-machine-learning/184006/5 and https://discuss.elastic.co/t/solved-cannot-start-es-7-0-1-on-fedora-30-unless-xpack-ml-enabled-false-is-set/179693 (but there have been many other posts like those over the years).

I think it's fine to start this code off in the `ml` plugin, especially as it's experimental.  But in the future I wonder if we should have a second `moar-ml` plugin that contains features that can work without the ML native code and have little/no need to share code that's already in the `ml` plugin.  `find_file_structure` is in the same situation and could also move across.

tveasey commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggregations/pca/PCAStatsResults.java:
```diff
@@ -0,0 +1,157 @@
+
+public class PCAStatsResults extends MatrixStatsResults {
+    /** eigen values from PCA */
+    protected final Map<String, Complex_F64> eigenVals;
```
The imaginary part should be identically zero for covariance matrix, it's Hermitian. I'd be tempted to store them as Double in eigenVals.

nknize replied:
:+1:

tveasey reviewed (commented):
I've done a first pass and made some suggestions. I've focused mainly on the PCA* classes and I'm not best placed to comment on the aspects of integrating with aggs framework.

nknize committed:
fix ci failures

nknize committed:
remove unused import

nknize committed:
fix equals hashcode

nknize committed:
pass precommit

nknize committed:
rebase to master

nknize commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggregations/pca/ParsedPCAStats.java:
```diff
@@ -0,0 +1,137 @@
+
+    private List<ParsedPCAStatsResults> eigenResults;
+    /** eigen values from PCA */
+    protected final Map<String, Complex_F64> eigenVals = new HashMap<>();
```
:+1:

nknize commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggregations/pca/PCAStatsResults.java:
```diff
@@ -0,0 +1,157 @@
+        }
+
+        // compute PCA as an eigendecomposition of the covariance matrix
+        EigenDecomposition_F64<DMatrixRMaj> decomp = DecompositionFactory_DDRM.eig(A.numRows, true);
```
:+1:

nknize commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggregations/pca/PCAStatsResults.java:
```diff
@@ -0,0 +1,157 @@
+
+        // compute PCA as an eigendecomposition of the covariance matrix
+        EigenDecomposition_F64<DMatrixRMaj> decomp = DecompositionFactory_DDRM.eig(A.numRows, true);
+        if( !decomp.decompose(A) ) {
```
:+1:

nknize commented on modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStats.java:
```diff
@@ -19,141 +19,54 @@
-
 /**
- * Computes distribution statistics over multiple fields
+ * Adds correlation computation to Computes distribution statistics over multiple fields
```
:+1:

nknize committed:
incorporate first round PR feedback

nknize commented:
Circled back around to this PR. Rebased with master and incorporated initial PR feedback. Next steps:

* refactor from ML to `analytics` plugin
* refactor to use Vector fields (looks like this will require a new `ValueType`?) /cc @not-napoleon @polyfractal  
* license Gold

nknize committed:
refactor PCA aggregation to analytics plugin

not-napoleon commented:
So, currently, Dense Vector Fields get built as a Bytes values source, which we generally treat to mean "String".  Once the registry goes live though, you're going to want to create a dedicated `ValuesSourceType` for it, otherwise you'll have to support anything that calls itself Bytes.  This also raises some BWC concerns, since right now Terms might work on vector fields (I haven't tried it, but in general terms works on bytes).  My work with values sources has led me to believe that a lot of plugin fields should have specialized values source types.

What's going to be tricky, either in pre or post registry implementation, is the fact that we're talking about cross-plugin communication without ever bouncing into core.  I.E., the vector field is in the vectors plugin and the PCA aggregation is in the analytics plugin (as opposed to what we've done so far which has been linking plugin fields to core aggregations).  This raises a bunch of questions, like "what if one or the other plugin isn't enabled?" and "what order do they load in?"   Currently, the registry keys aggregations by the `NAME` field from the builder; to register a new aggregator when adding a field, the registration call needs to know that name.  Which in this case means either hard coding it in two places or some kind of cross-plugin dependency.  Plus, we'll need some kind of cross-plugin dependency for testing.

If we don't want to solve the cross-plugin communication issue right now, we could instead move PCA into the vectors plugin.  With all the code in the same place, there's no loading issues or dependency issues.  That said, we'll need to solve this eventually, and this seems like an ideal test case.

nknize commented:
Thanks @not-napoleon This raises a great point. The PCA agg depends on `matrix_stats` foundation classes so this PR already introduces cross-plugin communication. Which means moving this to the `Vector` plugin won't help that issue. Probably a good idea to team-discuss these issues; I'll mark it for discussion.

droberts195 commented:
I don't think this is ever going to get merged in its current form.  Closing to avoid clutter in the ML PR list.
